# Creating Homework 1

In [1]:
import pandas as pd
from pprint import pprint
import seaborn as sns
import matplotlib.pylab as plt

%matplotlib inline

In [2]:
# read Country information from '../data/WDICountry.csv' into df_country
df_country = pd.read_csv('../data/WDICountry.csv', header=0)

In [3]:
# display the first 5 rows of df_country
df_country.head()

Country Code   Short Name   Table Name                     Long Name  \
0          ABW        Aruba        Aruba                         Aruba   
1          AFG  Afghanistan  Afghanistan  Islamic State of Afghanistan   
2          AGO       Angola       Angola   People's Republic of Angola   
3          ALB      Albania      Albania           Republic of Albania   
4          AND      Andorra      Andorra       Principality of Andorra   

  2-alpha code   Currency Unit  \
0           AW   Aruban florin   
1           AF  Afghan afghani   
2           AO  Angolan kwanza   
3           AL    Albanian lek   
4           AD            Euro   

                                       Special Notes  \
0  SNA data for 2000-2011 are updated from offici...   
1  Fiscal year end: March 20; reporting period fo...   
2                                                NaN   
3                                                NaN   
4  WB-3 code changed from ADO to AND to align wit...   

                      Region         Income Group WB-2 code     ...      \
0  Latin America & Caribbean          High income        AW     ...       
1                 South Asia           Low income        AF     ...       
2         Sub-Saharan Africa  Lower middle income        AO     ...       
3      Europe & Central Asia  Upper middle income        AL     ...       
4      Europe & Central Asia          High income        AD     ...       

                     IMF data dissemination standard  \
0                                                NaN   
1  Enhanced General Data Dissemination System (e-...   
2  Enhanced General Data Dissemination System (e-...   
3  Enhanced General Data Dissemination System (e-...   
4                                                NaN   

                            Latest population census  \
0                                               2010   
1                                               1979   
2                                               2014   
3                                               2011   
4  2011. Population data compiled from administra...   

                  Latest household survey  \
0                                     NaN   
1     Demographic and Health Survey, 2015   
2  Demographic and Health Survey, 2015/16   
3  Demographic and Health Survey, 2008/09   
4                                     NaN   

   Source of most recent Income and expenditure data  \
0                                                NaN   
1            Integrated household survey (IHS), 2011   
2         Integrated household survey (IHS), 2008/09   
3  Living Standards Measurement Study Survey (LSM...   
4                                                NaN   

  Vital registration complete Latest agricultural census  \
0                         Yes                        NaN   
1                         NaN                        NaN   
2                         NaN                        NaN   
3                         Yes                       2012   
4                         Yes                        NaN   

  Latest industrial data Latest trade data Latest water withdrawal data  \
0                    NaN            2015.0                          NaN   
1                    NaN            2015.0                       2000.0   
2                    NaN            2015.0                       2005.0   
3                 2013.0            2015.0                       2006.0   
4                    NaN            2014.0                          NaN   

  Unnamed: 31  
0         NaN  
1         NaN  
2         NaN  
3         NaN  
4         NaN  

[5 rows x 32 columns]

In [4]:
# display info for df_country
df_country.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 263 entries, 0 to 262
Data columns (total 32 columns):
Country Code                                         263 non-null object
Short Name                                           263 non-null object
Table Name                                           263 non-null object
Long Name                                            263 non-null object
2-alpha code                                         261 non-null object
Currency Unit                                        217 non-null object
Special Notes                                        150 non-null object
Region                                               217 non-null object
Income Group                                         217 non-null object
WB-2 code                                            262 non-null object
National accounts base year                          205 non-null object
National accounts reference year                     54 non-null float64
SNA price valuation

- How many rows are in this dataset?
263

- How many columns are in this dataset?
32

- How many null vales does column 'Region' have?
217

In [5]:
columns_to_keep = ['Country Code','Short Name','Region','Income Group']

In [6]:
# reduce df_country to only the column in columns_to_keep and display the DataFrame info
df_country = df_country[columns_to_keep] 
df_country.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 263 entries, 0 to 262
Data columns (total 4 columns):
Country Code    263 non-null object
Short Name      263 non-null object
Region          217 non-null object
Income Group    217 non-null object
dtypes: object(4)
memory usage: 8.3+ KB


- How many rows are now in this dataset?
263

- How many columns are now in this dataset?
4

- How many null vales does column 'Region' have? (hint: it shouldn't have changed)
217

In [7]:
# display the rows of df_country where 'Region' is null
df_country[df_country['Region'].isnull()]

Country Code                                         Short Name Region  \
5            ARB                                         Arab World    NaN   
34           CEB                     Central Europe and the Baltics    NaN   
47           CSS                             Caribbean small states    NaN   
59           EAP        East Asia & Pacific (excluding high income)    NaN   
60           EAR                         Early-demographic dividend    NaN   
61           EAS                                East Asia & Pacific    NaN   
62           ECA      Europe & Central Asia (excluding high income)    NaN   
63           ECS                              Europe & Central Asia    NaN   
66           EMU                                          Euro area    NaN   
71           EUU                                     European Union    NaN   
72           FCS           Fragile and conflict affected situations    NaN   
93           HIC                                        High income    NaN   
96           HPC             Heavily indebted poor countries (HIPC)    NaN   
100          IBD                                          IBRD only    NaN   
101          IBT                                   IDA & IBRD total    NaN   
102          IDA                                          IDA total    NaN   
103          IDB                                          IDA blend    NaN   
105          IDX                                           IDA only    NaN   
125          LAC  Latin America & Caribbean (excluding high income)    NaN   
131          LCN                          Latin America & Caribbean    NaN   
132          LDC       Least developed countries: UN classification    NaN   
133          LIC                                         Low income    NaN   
136          LMC                                Lower middle income    NaN   
137          LMY                                Low & middle income    NaN   
139          LTE                          Late-demographic dividend    NaN   
150          MEA                         Middle East & North Africa    NaN   
153          MIC                                      Middle income    NaN   
158          MNA  Middle East & North Africa (excluding high inc...    NaN   
167          NAC                                      North America    NaN   
178          OED                                       OECD members    NaN   
180          OSS                                 Other small states    NaN   
188          PRE                           Pre-demographic dividend    NaN   
194          PSS                        Pacific island small states    NaN   
195          PST                          Post-demographic dividend    NaN   
201          SAS                                         South Asia    NaN   
212          SSA         Sub-Saharan Africa (excluding high income)    NaN   
214          SSF                                 Sub-Saharan Africa    NaN   
215          SST                                       Small states    NaN   
227          TEA                   East Asia & Pacific (IDA & IBRD)    NaN   
228          TEC                 Europe & Central Asia (IDA & IBRD)    NaN   
233          TLA             Latin America & Caribbean (IDA & IBRD)    NaN   
235          TMN            Middle East & North Africa (IDA & IBRD)    NaN   
237          TSA                            South Asia (IDA & IBRD)    NaN   
238          TSS                    Sub-Saharan Africa (IDA & IBRD)    NaN   
246          UMC                                Upper middle income    NaN   
256          WLD                                              World    NaN   

    Income Group  
5            NaN  
34           NaN  
47           NaN  
59           NaN  
60           NaN  
61           NaN  
62           NaN  
63           NaN  
66           NaN  
71           NaN  
72           NaN  
93           NaN  
96           NaN  
100          NaN  
101          NaN  
102          NaN  
103          N

In [8]:
# drop the rows of df_country with null values inplace, and disply info for df_country
df_country.dropna(inplace=True)
df_country.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 217 entries, 0 to 262
Data columns (total 4 columns):
Country Code    217 non-null object
Short Name      217 non-null object
Region          217 non-null object
Income Group    217 non-null object
dtypes: object(4)
memory usage: 8.5+ KB


In [9]:
# assert that there are no duplicates (use len and drop_duplicates)
# hint: what two things should be the same length?
assert len(df_country.drop_duplicates()) == len(df_country)

In [10]:
# assert that 'Country Code' is unique (use len and unique)
# hist: again, what two things should be the same length?
assert len(df_country.loc[:,'Country Code'].unique()) == len(df_country)

In [11]:
# set the index of df_country to be 'Country Code' inplace and display the first 5 rows
df_country.set_index('Country Code', inplace=True)
df_country.head()

Short Name                     Region         Income Group
Country Code                                                             
ABW                 Aruba  Latin America & Caribbean          High income
AFG           Afghanistan                 South Asia           Low income
AGO                Angola         Sub-Saharan Africa  Lower middle income
ALB               Albania      Europe & Central Asia  Upper middle income
AND               Andorra      Europe & Central Asia          High income

In [12]:
# print the number of countries in df_country
n_countries = len(df_country)
print('n_countries: {}'.format(n_countries))

n_countries: 217


---

In [13]:
# read csv '../data/WDIData.csv' into df_data
df_data = pd.read_csv('../data/WDIData.csv')

In [14]:
# display the first 5 rows of df_data
df_data.head()

Country Name Country Code  \
0   Arab World          ARB   
1   Arab World          ARB   
2   Arab World          ARB   
3   Arab World          ARB   
4   Arab World          ARB   

                                      Indicator Name     Indicator Code  1960  \
0  2005 PPP conversion factor, GDP (LCU per inter...      PA.NUS.PPP.05   NaN   
1  2005 PPP conversion factor, private consumptio...  PA.NUS.PRVT.PP.05   NaN   
2  Access to clean fuels and technologies for coo...     EG.CFT.ACCS.ZS   NaN   
3            Access to electricity (% of population)     EG.ELC.ACCS.ZS   NaN   
4  Access to electricity, rural (% of rural popul...  EG.ELC.ACCS.RU.ZS   NaN   

   1961  1962  1963  1964  1965     ...            2008       2009       2010  \
0   NaN   NaN   NaN   NaN   NaN     ...             NaN        NaN        NaN   
1   NaN   NaN   NaN   NaN   NaN     ...             NaN        NaN        NaN   
2   NaN   NaN   NaN   NaN   NaN     ...       82.831932  83.474617  84.006076   
3   NaN   NaN   NaN   NaN   NaN     ...       85.042248  84.655338  85.955345   
4   NaN   NaN   NaN   NaN   NaN     ...       72.292323  71.765345  73.734266   

        2011       2012       2013       2014  2015  2016  Unnamed: 61  
0        NaN        NaN        NaN        NaN   NaN   NaN          NaN  
1        NaN        NaN        NaN        NaN   NaN   NaN          NaN  
2  84.416148  84.734574  85.003645  85.244971   NaN   NaN          NaN  
3  86.392313  86.846966  87.604961  88.039125   NaN   NaN          NaN  
4  74.500717  75.218497  76.712250  77.407269   NaN   NaN          NaN  

[5 rows x 62 columns]

In [15]:
# display info for df_data
df_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 401016 entries, 0 to 401015
Data columns (total 62 columns):
Country Name      401016 non-null object
Country Code      401016 non-null object
Indicator Name    401016 non-null object
Indicator Code    401016 non-null object
1960              34646 non-null float64
1961              38504 non-null float64
1962              40438 non-null float64
1963              40738 non-null float64
1964              41260 non-null float64
1965              44146 non-null float64
1966              43808 non-null float64
1967              45298 non-null float64
1968              45143 non-null float64
1969              46858 non-null float64
1970              79998 non-null float64
1971              86577 non-null float64
1972              89660 non-null float64
1973              89423 non-null float64
1974              90945 non-null float64
1975              95139 non-null float64
1976              97554 non-null float64
1977              101394 non

In [17]:
# pprint a list of the unique values in Indicator Name
pprint(list(df_data.loc[:,'Indicator Name'].unique()))

['2005 PPP conversion factor, GDP (LCU per international $)',
 '2005 PPP conversion factor, private consumption (LCU per international $)',
 'Access to clean fuels and technologies for cooking  (% of population)',
 'Access to electricity (% of population)',
 'Access to electricity, rural (% of rural population)',
 'Access to electricity, urban (% of urban population)',
 'Account (% age 15+) [ts]',
 'Account at a financial institution (% age 15+)',
 'Account at a financial institution, female (% age 15+)',
 'Account at a financial institution, income, poorest 40% (% ages 15+)',
 'Account at a financial institution, income, richest 60% (% ages 15+)',
 'Account at a financial institution, male (% age 15+)',
 'Account, female (% age 15+) [ts]',
 'Account, income, poorest 40% (% ages 15+) [ts]',
 'Account, income, richest 60% (% ages 15+) [ts]',
 'Account, male (% age 15+) [ts]',
 'Account, older adults (% ages 25+) [ts]',
 'Account, primary education or less (% ages 15+) [ts]',
 'Account, 

 '100)',
 'Permanent cropland (% of land area)',
 'Persistence to grade 5, female (% of cohort)',
 'Persistence to grade 5, male (% of cohort)',
 'Persistence to grade 5, total (% of cohort)',
 'Persistence to last grade of primary, female (% of cohort)',
 'Persistence to last grade of primary, male (% of cohort)',
 'Persistence to last grade of primary, total (% of cohort)',
 'Personal remittances, paid (current US$)',
 'Personal remittances, received (% of GDP)',
 'Personal remittances, received (current US$)',
 'Personal transfers, receipts (BoP, current US$)',
 'PFC gas emissions (thousand metric tons of CO2 equivalent)',
 'Physicians (per 1,000 people)',
 'Plant species (higher), threatened',
 'PM2.5 air pollution, mean annual exposure (micrograms per cubic meter)',
 'PM2.5 air pollution, population exposed to levels exceeding WHO guideline '
 'value (% of total)',
 'PNG, bonds (AMT, current US$)',
 'PNG, bonds (DIS, current US$)',
 'PNG, bonds (DOD, current US$)',
 'PNG, bonds (I

In [18]:
# these are the indicators we will keep
data_indicators_to_keep = ['Access to electricity (% of population)','GDP (constant 2010 US$)','Population, total','Population density (people per sq. km of land area)','Unemployment, total (% of total labor force) (national estimate)']

In [19]:
# these are the columns we will keep
data_columns_to_keep = ['Country Code','Indicator Name','2014']

In [23]:
# reduce data to only 
#  the rows whose 'Indicator Name' isin data_indicators_to_keep and 
#  the columns in data_columns_to_keep
df_data = df_data.loc[df_data['Indicator Name'].isin(data_indicators_to_keep),data_columns_to_keep]

In [24]:
# display the info for df_data
df_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1320 entries, 3 to 400977
Data columns (total 3 columns):
Country Code      1320 non-null object
Indicator Name    1320 non-null object
2014              1166 non-null float64
dtypes: float64(1), object(2)
memory usage: 41.2+ KB


In [26]:
# into df_data_pivot, pivot df_data with index 'Country Code', columns 'Indicator Name' and values '2014'
# and display the first 5 rows
df_data_pivot = df_data.pivot(index='Country Code',columns='Indicator Name',values='2014')
df_data_pivot.head()

Indicator Name  Access to electricity (% of population)  \
Country Code                                              
ABW                                           94.468323   
AFG                                           89.500000   
AGO                                           32.000000   
ALB                                          100.000000   
AND                                          100.000000   

Indicator Name  GDP (constant 2010 US$)  \
Country Code                              
ABW                                 NaN   
AFG                        1.999032e+10   
AGO                        1.008863e+11   
ALB                        1.276823e+10   
AND                                 NaN   

Indicator Name  Population density (people per sq. km of land area)  \
Country Code                                                          
ABW                                                    576.638889     
AFG                                                     50.176179     
AGO                                                     21.593379     
ALB                                                    105.441752     
AND                                                    168.559574     

Indicator Name  Population, total  \
Country Code                        
ABW                      103795.0   
AFG                    32758020.0   
AGO                    26920466.0   
ALB                     2889104.0   
AND                       79223.0   

Indicator Name  Unemployment, total (% of total labor force) (national estimate)  
Country Code                                                                      
ABW                                                           NaN                 
AFG                                                           NaN                 
AGO                                                           NaN                 
ALB                                                         17.49                 
AND                                                           NaN

In [27]:
# display info for df_data_pivot
df_data_pivot.info()

<class 'pandas.core.frame.DataFrame'>
Index: 264 entries, ABW to ZWE
Data columns (total 5 columns):
Access to electricity (% of population)                             260 non-null float64
GDP (constant 2010 US$)                                             239 non-null float64
Population density (people per sq. km of land area)                 261 non-null float64
Population, total                                                   262 non-null float64
Unemployment, total (% of total labor force) (national estimate)    144 non-null float64
dtypes: float64(5)
memory usage: 12.4+ KB


In [32]:
# set df as the inner join df_country with df_data_pivot and display the first 5 rows
df = df_country.join(df_data_pivot, how='inner')
df.head()

Short Name                     Region         Income Group  \
Country Code                                                                
ABW                 Aruba  Latin America & Caribbean          High income   
AFG           Afghanistan                 South Asia           Low income   
AGO                Angola         Sub-Saharan Africa  Lower middle income   
ALB               Albania      Europe & Central Asia  Upper middle income   
AND               Andorra      Europe & Central Asia          High income   

              Access to electricity (% of population)  \
Country Code                                            
ABW                                         94.468323   
AFG                                         89.500000   
AGO                                         32.000000   
ALB                                        100.000000   
AND                                        100.000000   

              GDP (constant 2010 US$)  \
Country Code                            
ABW                               NaN   
AFG                      1.999032e+10   
AGO                      1.008863e+11   
ALB                      1.276823e+10   
AND                               NaN   

              Population density (people per sq. km of land area)  \
Country Code                                                        
ABW                                                  576.638889     
AFG                                                   50.176179     
AGO                                                   21.593379     
ALB                                                  105.441752     
AND                                                  168.559574     

              Population, total  \
Country Code                      
ABW                    103795.0   
AFG                  32758020.0   
AGO                  26920466.0   
ALB                   2889104.0   
AND                     79223.0   

              Unemployment, total (% of total labor force) (national estimate)  
Country Code                                                                    
ABW                                                         NaN                 
AFG                                                         NaN                 
AGO                                                         NaN                 
ALB                                                       17.49                 
AND                                                         NaN

In [34]:
# display info for df (are there the right number of rows, the same as number of countries?)
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 217 entries, ABW to ZWE
Data columns (total 8 columns):
Short Name                                                          217 non-null object
Region                                                              217 non-null object
Income Group                                                        217 non-null object
Access to electricity (% of population)                             214 non-null float64
GDP (constant 2010 US$)                                             193 non-null float64
Population density (people per sq. km of land area)                 215 non-null float64
Population, total                                                   216 non-null float64
Unemployment, total (% of total labor force) (national estimate)    113 non-null float64
dtypes: float64(5), object(3)
memory usage: 15.3+ KB


In [36]:
# display the value counts for Region
df['Region'].value_counts()

Europe & Central Asia         58
Sub-Saharan Africa            48
Latin America & Caribbean     42
East Asia & Pacific           37
Middle East & North Africa    21
South Asia                     8
North America                  3
Name: Region, dtype: int64

In [40]:
# into df_region_dummies, create a dummies matrix for region, dropping the first region value
# and display the first 5 rows
df_region_dummies = pd.get_dummies(df['Region'], drop_first=True)
df_region_dummies.head()

Europe & Central Asia  Latin America & Caribbean  \
Country Code                                                     
ABW                               0                          1   
AFG                               0                          0   
AGO                               0                          0   
ALB                               1                          0   
AND                               1                          0   

              Middle East & North Africa  North America  South Asia  \
Country Code                                                          
ABW                                    0              0           0   
AFG                                    0              0           1   
AGO                                    0              0           0   
ALB                                    0              0           0   
AND                                    0              0           0   

              Sub-Saharan Africa  
Country Code                      
ABW                            0  
AFG                            0  
AGO                            1  
ALB                            0  
AND                            0

Which region was dropped?
East Asia & Pacific

In [ ]:
g = df.groupby('Region')

In [ ]:
g['Access to electricity (% of population)'].mean()

In [ ]:
g['Access to electricity (% of population)'].median()

In [ ]:
sns.boxplot(x='Region',y='Access to electricity (% of population)',data=df)
_ = plt.xticks(rotation=90)

In [ ]:
df